rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+
+- One part contains independent<sup>*</sup> resources and is defined within
+the IaC application (AWS CDK or CDKTF).
+- The second part contains the resources being used by Kubernetes resources,
+and are provisioned imperatively before synthesis, either as part of the
+cdk8s application, or externally. The imperative nature of provisioning allows waiting
+on resource creation and passing any required information to downstream processes, 
+such as `cdk8s synth`.
+
+> <sup>*</sup> Independent with respect to usage in Kubernetes resource definitions.
+
+However, this separation into imperative provisioning essentially re-introduces all
+the complexity that IaC aims to solve, and is not desirable.
+
+#### Post Synth Processing
+
+In this approach, cloud infrastructure and Kubernetes resources are defined in the same
+application using the standard IaC tooling. The Kubernetes resources are tightly coupled
+with their required cloud resources. 
+
+```ts
+...
+
+// define the bucket without an explicit name
+new aws.aws_s3.Bucket(stack, 'Bucket');
+
+// pass the bucket name to the CronJob container
+new kplus.CronJob(manifest, 'CronJob', {
+ containers: [{ 
+   image: 'job',
+   envVariables: {
+     BUCKET_NAME: kplus.EnvValue.fromValue(bucket.bucketName),
+   }
+ }]
+});
+
+cdk8sApp.synth();
+
+...
+```
+
+To overcome the problem of unresolved tokens, customers have to post-process the result 
+of `cdk8sApp.synth()` to produce the final deployable manifest. This post synthesis 
+step inspects the manifest for references to cloud resources, interprets them, 
+and performs the necessary lookups.
+
+From an application maintenance perspective, this approach is actually pretty robust.
+The problem is that implementing and maintaining such a post synthesis step is not at all
+trivial. This RFC essentially proposes baking this step into the cdk8s framework, so that
+customers don't have to deal with the complexities it poses.
+
+### What is the technical solution (design) of this feature?
+
+The high-level design consists of three parts:
+
+#### Custom Resolvers
+
+When cdk8s applications are synthesized, the [`resolve`](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L3) 
+function is called for every property of the user defined resource spec. Currently, it handles 
+resolving instances of [`Lazy`](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L10) 
+and [implicit tokens](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L16).
+
+To allow the lookups described in this RFC, we will provide a custom resolution hook in this 
+function in the form of an interface, that customers can implement:
+
+```ts
+import { ApiObject } from './api-object';
+
+export class ResolutionContext {
+
+  public readonly obj: ApiObject;
+  public readonly key: string[];
+
+}
+
+export interface IResolver {
+
+  resolve(context: ResolutionContext, value: any): any;
```

Comment:
For future extensibility, I think I would offer an object with methods. Rather than:

```ts
if (shouldDoSomething(value)) {
  return doSomething(newValue); 
}
return value;
```

The implement would look like:

```ts
if (shouldDoSomething(value)) {
  context.replace(newValue); 
}
```

And future operations or tweaks could be made by extending the set of possible operations on `context`:

```ts
if (shouldDoSomething(value)) {
  context.replace(newValue, { cache: true }); 
}
```

Reply:
Done 👍